### PR TITLE
WIP: Fix/issue 251

### PR DESCRIPTION
**Pull request by Sirius207**
## Description
Follow the discussion of #251, I add the following command. 
```
$ cz undo -h
$ cz undo --bump
$ cz undo --commit
```
## Checklist
- [x] Add test cases to all the changes you introduce
- [x] Run `./script/format` and `./script/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes
## Expected behavior

`cz undo --bump`: delete the latest tag and the related commit
`cz undo --commit`: reset the latest commit
## Steps to Test This Pull Request
`cz undo --bump`
1. run `cz bump`
2. run `cz undo --bump`
`cz undo --commit`
1. run `cz commit`
2. run `cz undo --commit`
## Additional context

If I execute the following statements first,  the `git.get_tags()` would get the tags order by `[0.3.0a0, 0.3.0, ...]`.
image
image
image
So I implement another function `git.get_latest_tag()`

### CODE REVIEW

**Comment by Woile** [comment]
Nice, thanks for this :+1: 

Could you include a new command documentation in `docs/undo.md` and update `mkdocs.yml` with it? Thanks

File: commitizen/cli.py [hunk 1]
```diff
@@ -133,6 +133,23 @@
                     },
                 ],
             },
+            {
+                "name": "undo",
+                "help": "revert bump or commit",
+                "func": commands.Undo,
+                "arguments": [
+                    {
+                        "name": ["--bump"],
```
Comment (Woile) [inline commitizen/cli.py:142]: What do you think of adding a shortcut here? `-b` and `-c` for commit?
We'd have `cz undo -b`

File: commitizen/commands/undo.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        if created_tag.rev != created_commit.rev:
+            raise InvalidCommandArgumentError(
+                "The revision of the latest tag is not equal to the latest commit, use git undo --commit instead\n\n"
+                f"Latest Tag: {created_tag.name}, {created_tag.rev}, {created_tag.date}\n"
+                f"Latest Commit: {created_commit.title}, {created_commit.rev}"
+            )
+
+        command = f"git tag --delete {created_tag.name} && git reset HEAD~ && git reset --hard HEAD"
+
```
Comment (Woile) [inline commitizen/commands/undo.py:29]: I would print a message here:

```python
out.info("Reverting version bump, running:")
out.info(f"{command}")
out.info(f"The tag can be removed from a remote by running `git push origin :{created_tag.name}`
```

I would print the command to educate the users and show what's actually happening, not much magic.

**Comment by Sirius207** [comment]
> Nice, thanks for this 👍
> 
> Could you include a new command documentation in `docs/undo.md` and update `mkdocs.yml` with it? Thanks
> 
> We are working on fixing the CI issue.

Sure~ 👍

File: commitizen/cli.py (hunk 1 above)
Comment (Sirius207) [reply commitizen/cli.py:142]: Sounds great 💯

**Comment by Lee-W** [comment]
@Sirius207 Even though we're working on the CI issue, would you help us ignore `commitizen/commands/init.py:50` and fix `commitizen/commands/init.py:82`? You can ignore line 50 by adding `# type: ignore` after that line. As for fixing line 82, you'll have to change `commitizen/git.py:126` from ` Optional[List[str]]` to `List[Optional[str]]` and you'll still have to add `# type: ignore` after line 82.

**Comment by Sirius207** [comment]
> @Sirius207 Even though we're working on the CI issue, would you help us ignore `commitizen/commands/init.py:50` and fix `commitizen/commands/init.py:82`? You can ignore line 50 by adding `# type: ignore` after that line. As for fixing line 82, you'll have to change `commitizen/git.py:126` from ` Optional[List[str]]` to `List[Optional[str]]` and you'll still have to add `# type: ignore` after line 82.

OK 👍

File: commitizen/commands/undo.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+from commitizen import cmd, factory, git, out
+from commitizen.config import BaseConfig
+from commitizen.exceptions import InvalidCommandArgumentError
+
+
+class Undo:
+    """Reset the latest git commit or git tag."""
```
Comment (Lee-W) [inline commitizen/commands/undo.py:7]: It seems we use various terms for the same meaning (i.e., reset / revert / undo). I'll suggest unifying them.

File: commitizen/commands/undo.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    def __init__(self, config: BaseConfig, arguments: dict):
+        self.config: BaseConfig = config
+        self.cz = factory.commiter_factory(self.config)
+        self.arguments = arguments
+
+    def _get_bump_command(self):
+        created_tag = git.get_latest_tag()
```
Comment (Lee-W) [inline commitizen/commands/undo.py:15]: Why is this one named as `created_tag` instead of `latest_tag`?

File: commitizen/commands/undo.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        commits = git.get_commits()
+
+        if created_tag and commits:
+            created_commit = commits[0]
```
Comment (Lee-W) [inline commitizen/commands/undo.py:19]: same as the question above

File: commitizen/commands/undo.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        out.info("Reverting version bump, running:")
+        out.info(f"{command}")
+        out.info(
+            f"The tag can be removed from a remote by running `git push origin :{created_tag.name}`"
+        )
+
+        return command
+
+    def __call__(self):
+        bump: bool = self.arguments.get("bump")
```
Comment (Lee-W) [inline commitizen/commands/undo.py:41]: `bump` doesn't seem to be clear enough to me. How about something like `reverting_bump`?

File: commitizen/commands/undo.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        commit: bool = self.arguments.get("commit")
+
+        if bump:
+            command = self._get_bump_command()
+        elif commit:
+            command = "git reset HEAD~"
```
Comment (Lee-W) [inline commitizen/commands/undo.py:47]: Sometimes we'll encounter the situation that we don't have `HEAD`. Will it be possible for us to solve that situation?

File: commitizen/commands/undo.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                    "See 'cz undo -h' for more information"
+                )
+            )
+
+        c = cmd.run(command)
+        if c.err:
+            out.error(c.err)
+
+        out.write(c.out)
+        out.success("Undo successful!")
```
Comment (Lee-W) [inline commitizen/commands/undo.py:61]: Maybe `Succeeded on undoning!` or `Undo successfully!`?

File: commitizen/commands/undo.py (hunk 3 above)
Comment (Sirius207) [reply commitizen/commands/undo.py:7]: I could use "undo". 👌

File: commitizen/commands/undo.py (hunk 4 above)
Comment (Sirius207) [reply commitizen/commands/undo.py:15]: The `created_tag` is original name and `get_latest_tag` comes later, I would use `latest_tag` in next commit.

File: commitizen/commands/undo.py (hunk 8 above)
Comment (Sirius207) [reply commitizen/commands/undo.py:61]: I prefer the latter 👍

**Comment by Woile** [comment]
Will use documentation for now, closing this

**Pull request closed without merging**